### PR TITLE
GH-43966: [Java] Check for nullabilities when comparing StructVector

### PR DESCRIPTION
**Pull request by hellishfire**
### Rationale for this change

See #43966

### What changes are included in this PR?

Check for nullabilities when comparing StructVector with RangeEqualsVisitor.

### Are these changes tested?

Yes

### Are there any user-facing changes?
No

* GitHub Issue: #43966

### CODE REVIEW

**Comment by vibhatha** [comment]
@hellishfire  let's make the PR title as same as the issue title.

File: java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java [hunk 1]
```diff
@@ -354,6 +355,18 @@ protected boolean compareStructVectors(Range range) {
       return false;
     }
 
+    if (leftVector instanceof StructVector || rightVector instanceof StructVector) {
```
Comment (vibhatha) [inline java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java:358]: What if we do this check for all vectors not just `StructVector`?

Comment (vibhatha) [reply java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java:358]: Another thing is this visitor interface is used for all vectors, and specifying a logic just for a single vector type doesn't make much sense. Further, if we are to check the range considering nulls, it should be for all vectors?

Comment (hellishfire) [reply java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java:358]: There's already null compacrion for other types of vector.

Comment (vibhatha) [reply java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java:358]: Oh, I misinterpreted.

**Comment by hellishfire** [comment]
> @hellishfire let's make the PR title as same as the issue title.

I renamed issue title to be the same as PR title

**Review by vibhatha — APPROVED** [review]
Thanks for making this PR.

**Comment by vibhatha** [comment]
@lidavidm appreciate your review and also please approve the workflows.

**Review by lidavidm — COMMENTED** [review]
Do null slots still compare equal when the child values are different? I'd like to see a unit test for that. (They should, but it looks like we just directly dispatch to RangeEqualsVisitor.)

**Comment by hellishfire** [comment]
> Do null slots still compare equal when the child values are different? I'd like to see a unit test for that. (They should, but it looks like we just directly dispatch to RangeEqualsVisitor.)

Currently not. Further code change is needed to achieve this.

**Comment by vibhatha** [comment]
Probably we should address that as well. I think it is in scope. Sorry I missed that.

**Comment by hellishfire** [comment]
@vibhatha @lidavidm 
I rewrote the compare logic and related tests

File: java/vector/src/test/java/org/apache/arrow/vector/compare/TestRangeEqualsVisitor.java [hunk 2]
```diff
@@ -434,25 +434,39 @@ public void testStructVectorRangeEquals() {
...
       writer1.allocate();
 
+      writeStructVector(writer1, 0, 0L);
       writeStructVector(writer1, 1, 10L);
       writeStructVector(writer1, 2, 20L);
       writeStructVector(writer1, 3, 30L);
       writeStructVector(writer1, 4, 40L);
       writeStructVector(writer1, 5, 50L);
-      writer1.setValueCount(5);
+      writer1.setValueCount(6);
```
Comment (vibhatha) [inline java/vector/src/test/java/org/apache/arrow/vector/compare/TestRangeEqualsVisitor.java:443]: We are adding 6 values? And why we set count to 5?

File: java/vector/src/test/java/org/apache/arrow/vector/compare/TestRangeEqualsVisitor.java [hunk 3]
```diff
@@ -434,25 +434,39 @@ public void testStructVectorRangeEquals() {
...
+      writeStructVector(writer2, 0, 0L);
       writeStructVector(writer2, 2, 20L);
       writeStructVector(writer2, 3, 30L);
       writeStructVector(writer2, 4, 40L);
       writeStructVector(writer2, 0, 0L);
       writer2.setValueCount(5);
 
       RangeEqualsVisitor visitor = new RangeEqualsVisitor(vector1, vector2);
-      assertTrue(visitor.rangeEquals(new Range(1, 1, 3)));
+      assertTrue(visitor.rangeEquals(new Range(2, 1, 3)));
```
Comment (vibhatha) [inline java/vector/src/test/java/org/apache/arrow/vector/compare/TestRangeEqualsVisitor.java:456]: Maybe write a new test to evaluate the new case?

File: java/vector/src/test/java/org/apache/arrow/vector/compare/TestRangeEqualsVisitor.java (hunk 2 above)
Comment (hellishfire) [reply java/vector/src/test/java/org/apache/arrow/vector/compare/TestRangeEqualsVisitor.java:443]: I just want to test the case where the left&right index in the range are different.
This is better than the original case where the left & right index are identical.

**Review by lidavidm — APPROVED** [review]

**Pull request merged**